xarray/coding/times.py

```diff
@@ -579,20 +579,23 @@ def _numbers_to_timedelta(
 ) -> np.ndarray:
     """Transform numbers to np.timedelta64."""
     # keep NaT/nan mask
-    nan = np.isnan(flat_num) | (flat_num == np.iinfo(np.int64).min)
+    if flat_num.dtype.kind == "f":
+        nan = np.asarray(np.isnan(flat_num))
+    elif flat_num.dtype.kind == "i":
+        nan = np.asarray(flat_num == np.iinfo(np.int64).min)
 
     # in case we need to change the unit, we fix the numbers here
     # this should be safe, as errors would have been raised above
     ns_time_unit = _NS_PER_TIME_DELTA[time_unit]
     ns_ref_date_unit = _NS_PER_TIME_DELTA[ref_unit]
     if ns_time_unit > ns_ref_date_unit:
-        flat_num *= np.int64(ns_time_unit / ns_ref_date_unit)
+        flat_num = np.asarray(flat_num * np.int64(ns_time_unit / ns_ref_date_unit))
```

Comment:
For pandas 3.0 CoW (copy on write) this array might be flagged readonly. So creating a copy here. The asarray-wrapping is needed for upstream-dev to keep 0d-arrays an array.